<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • rPHG2</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.9/font.css" rel="stylesheet"><link href="../deps/Roboto_Mono-0.4.9/font.css" rel="stylesheet"><link href="../deps/Oswald-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rPHG2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rPHG2.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/metrics.html">Load and visualize PHG metrics</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rPHG2/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/phg_v2" aria-label="PHG Source"><span class="fa fa-dna"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/maize-genetics/rPHG2/blob/main/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8" id="rphg2-08">rPHG2 0.8<a class="anchor" aria-label="anchor" href="#rphg2-08"></a></h2>
<ul><li>Fixed missing IDs returned in <code><a href="../reference/readHapIds.html">readHapIds()</a></code> method calls:
<ul><li>missing values (i.e., <code>"null"</code>) now return as <code>NA</code>s in the matrix object</li>
</ul></li>
<li>Removed <code><a href="https://rlang.r-lib.org/reference/abort.html" class="external-link">rlang::abort()</a></code> condition in <code><a href="../reference/initPhg.html">initPhg()</a></code>:
<ul><li>Now returns “warning” message instead of low-level <code><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop()</a></code> procedure</li>
</ul></li>
<li>Modified the <code><a href="../reference/PHGLocalCon.html">PHGLocalCon()</a></code> constructor to now take in either files or a directory path
<ul><li>Future plans for integrating TileDB and this release’s AGC interface</li>
</ul></li>
<li>Fixed plotting bug in <code><a href="../reference/plotHaploCounts.html">plotHaploCounts()</a></code> <code>geom</code> parameter</li>
<li>Added new <code>read*</code> function, <code><a href="../reference/readSequence.html">readSequence()</a></code>
<ul><li>Returns memory-efficient sequence data from an AGC instance as a <code><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">Biostrings::DNAString</a></code> object</li>
<li>Can return singlular haplotype sequences or entire reference ranges of sequence data</li>
<li>Allows for upstream and downstream padding via the <code>pad</code> parameter</li>
<li>This function suite also provides methods of linking a PHGv2 Conda environment to your global R options via the <code><a href="../reference/linkCondaBin.html">linkCondaBin()</a></code> function</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7" id="rphg2-07">rPHG2 0.7<a class="anchor" aria-label="anchor" href="#rphg2-07"></a></h2>
<ul><li>Added new parameter to <code><a href="../reference/plotHaploCounts.html">plotHaploCounts()</a></code>:
<ul><li><code>geom</code></li>
<li>allows user to specify a given geometry (line, bar, or point)</li>
</ul></li>
<li>Added new conditionals for to <code><a href="../reference/buildHaplotypeGraph.html">buildHaplotypeGraph()</a></code>:
<ul><li>Checks to see if the JVM is initialized. If not, it will cause an exception and direct user to run <code><a href="../reference/initPhg.html">initPhg()</a></code>
</li>
</ul></li>
<li>Added new class <code>JvmStats</code>:
<ul><li>Returns internal Java and PHG versions and memory information for JVM session</li>
<li>Can be instantiated using the <code><a href="../reference/jvmStats.html">jvmStats()</a></code> constructor function</li>
</ul></li>
<li>Added new checks for <code><a href="../reference/initPhg.html">initPhg()</a></code>
</li>
<li>Added new parameters to <code><a href="../reference/plotGvcf.html">plotGvcf()</a></code> function:
<ul><li>
<code>colorOverride</code>: Overrides bar color based on given user defined color</li>
<li>
<code>sampleOrder</code>: Overrides default bar ordering based on user defined <code>character</code> vector</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6" id="rphg2-06">rPHG2 0.6<a class="anchor" aria-label="anchor" href="#rphg2-06"></a></h2>
<ul><li>Added new parameters to <code>plogtGvcf()</code>:
<ul><li>
<code>mData</code> and <code>mVar</code>
</li>
<li>Allows user to override default sample color option with categorical data</li>
</ul></li>
<li>Added new generic, <code><a href="../reference/numberOfHaplotypes.html">numberOfHaplotypes()</a></code>:
<ul><li>Returns the number of unique haplotypes found within each reference range or the total value found within a <code>PHGDataSet</code> object</li>
</ul></li>
<li>Added getters to <code>PHGDataSet</code> objects:
<ul><li><code><a href="../reference/numberOfChromosomes.html">numberOfChromosomes()</a></code></li>
<li><code><a href="../reference/numberOfSamples.html">numberOfSamples()</a></code></li>
<li><code><a href="../reference/numberOfRefRanges.html">numberOfRefRanges()</a></code></li>
</ul></li>
<li>Changed functions and variables containing <code>taxa</code> to <code>samples</code>
<ul><li>e.g., <code>numberOfTaxa()</code> is now <code><a href="../reference/numberOfSamples.html">numberOfSamples()</a></code>
</li>
</ul></li>
<li>Format tweaks to <code>HaplotypeGraph</code> objects’ <code>show()</code> method</li>
<li>Added new function <code>plotHaploCount()</code>:
<ul><li>Plots unique haplotype counts for all or given reference ranges</li>
</ul></li>
<li>Added new function <code><a href="../reference/plotHaploDist.html">plotHaploDist()</a></code>:
<ul><li>Plots unique haplotype distribution for entire <code>PHGDataSet</code> object</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5" id="rphg2-05">rPHG2 0.5<a class="anchor" aria-label="anchor" href="#rphg2-05"></a></h2>
<ul><li>Added new visualization method, <code><a href="../reference/plotGvcf.html">plotGvcf()</a></code>:
<ul><li>Auto plotting various gVCF metrics</li>
<li>Granular metric support through formula subsetting</li>
</ul></li>
<li>Added new accessor and setting method, <code>seqnames()</code>
<ul><li>Returns all contig IDs found in a <code>PHGMetrics</code> object</li>
<li>Setter version (<code>seqnames()&lt;-</code>) will update old IDs found within a <code>data.frame</code> object</li>
</ul></li>
<li>Added new default methods to <code><a href="../reference/plotGvcf.html">plotGvcf()</a></code> and <code><a href="../reference/plotDot.html">plotDot()</a></code> for <code>PHGMetrics</code> objects that only contain one alignment or gVCF dataset</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4" id="rphg2-04">rPHG2 0.4<a class="anchor" aria-label="anchor" href="#rphg2-04"></a></h2>
<ul><li>Added vignettes and README updates
<ul><li><code><a href="../articles/rPHG2.html">vignette("rPHG2")</a></code></li>
</ul></li>
<li>Created build system for website generation
<ul><li><a href="https://rphg2.maizegenetics.net">rphg2.maizegenetics.net</a></li>
</ul></li>
<li>Added new class <code>PHGMetrics</code>
<ul><li>Primary container for PHGv2 metrics data</li>
</ul></li>
<li>Added accessor methods for <code>PHGMetrics</code> objects:
<ul><li><code><a href="../reference/metricsIds.html">metricsIds()</a></code></li>
<li><code><a href="../reference/metricsTable.html">metricsTable()</a></code></li>
<li><code><a href="../reference/metricsMetaData.html">metricsMetaData()</a></code></li>
</ul></li>
<li>Prior methods are also updatable</li>
<li>Added visualization methods:
<ul><li>
<code><a href="../reference/plotDot.html">plotDot()</a></code> - collinearity plotter for alignment data</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3" id="rphg2-03">rPHG2 0.3<a class="anchor" aria-label="anchor" href="#rphg2-03"></a></h2>
<ul><li>Added new class, <code>PHGDataSet</code>
<ul><li>In memory representation of hVCF data sets</li>
</ul></li>
<li>Added new function, <code><a href="../reference/filterRefRanges.html">filterRefRanges()</a></code>
<ul><li>Filters relevant rPHG2 data objects by reference range</li>
<li>Requires a <code>GRanges</code> object as input</li>
</ul></li>
<li>Added new function, <code><a href="../reference/filterSamples.html">filterSamples()</a></code>
<ul><li>Filters relevant rPHG2 data objects by sample ID</li>
<li>Input is a basic <code>character</code> vector of sample identifiers</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2" id="rphg2-02">rPHG2 0.2<a class="anchor" aria-label="anchor" href="#rphg2-02"></a></h2>
<ul><li>Added JVM connections via <code>rJava</code> interface</li>
<li>Added new class, <code>HaplotypeGraph</code>
<ul><li>Wrapper for PHGv2 API Java graph object</li>
</ul></li>
<li>Added new class, <code>PHGLocalCon</code>
<ul><li>Interface to local hVCF database and file connections</li>
</ul></li>
<li>Added new function <code>initPHG()</code>
<ul><li>Initializes JVM and adds PHGv2 Java JARs to classpath</li>
</ul></li>
<li>Added new function <code><a href="../reference/readHapIds.html">readHapIds()</a></code>
<ul><li>Reads haplotype IDs as a <code>character</code> matrix from a connection object</li>
</ul></li>
<li>Added new function <code><a href="../reference/readHapIdMetaData.html">readHapIdMetaData()</a></code>
<ul><li>Reads ALT header metadata for each haplotype ID from a connection object</li>
</ul></li>
<li>Added new function <code><a href="../reference/readHapIdPosMetaData.html">readHapIdPosMetaData()</a></code>
<ul><li>Reads ALT header positional metadata for each haplotype ID from a connection object</li>
</ul></li>
<li>Added new function <code><a href="../reference/readSamples.html">readSamples()</a></code>
<ul><li>Reads sample IDs from a connection object</li>
</ul></li>
<li>Added new function <code><a href="../reference/readRefRanges.html">readRefRanges()</a></code>
<ul><li>Reads reference range positional information as a <code>GRanges</code> object from a connection object</li>
</ul></li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p><b>rPHG2</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

