---
title: "Read haplotype sequence information"
output: rmarkdown::html_vignette
description: >
  Learn how to read sequence data from AGC files in PHG databases
vignette: >
  %\VignetteIndexEntry{Load and visualize PHG metrics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
editor_options:
  markdown:
    wrap: 70
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.align = "center",
  external  = TRUE,
  echo      = TRUE,
  warning   = FALSE
)

library(rPHG2)

source("../tests/testthat/setup.R")

options("phgv2_agc_path" = agcVanillaPath)

hVcfFiles <- system.file(
    "extdata", 
    c("LineA.h.vcf", "LineB.h.vcf"), 
    package = "rPHG2"
)

phgDs <- PHGLocalCon(hVcfFiles, dbUri = phgLibDir) |> 
    buildHaplotypeGraph() |>
    readPhgDataSet()
```

## Returning sequence data

We can also return haplotype sequence information via the `readSequence()`
command with PHG data sets (i.e., `PHGDataSet` objects) pointed to [PHGv2 databases
containing compressed assemblies information](https://phg.maizegenetics.net/build_and_load/#compress-fasta-files).

In order to perform sequence retrieval, we must perform a couple of steps:

1. Add `host` information to a `PHGLocalCon` object
2. Add an AGC path to the global options (i.e., the `options()` function)



### Setup - add host information
For **step 1**, during the the construction of a `PHGLocalCon` object using
the `PHGLocalCon()` function, add the path to a 
[PHGv2 database](https://phg.maizegenetics.net/build_and_load/#initialize-tiledb-instances) 
containing a compressed sequence file: 

```{r, echo=TRUE, eval=FALSE}
dbPath <- "/my/phgv2/db_path/"

# using 'hVcfFiles' variable from "Getting Started" steps
localCon <- hVcfFiles |> PHGLocalCon(dbUri = dbPath)
```

If you are unsure what this path is, it should be the directory that was
created during the PHGv2 building and loading steps and contains files and
sub-directories that looks like the following:

```
phg_v2_example/
├── data
│   ├── anchors.gff
│   ├── annotation_keyfile.txt
│   ├── Ref-v5.fa
│   ├── LineA-final-01.fa
│   └── LineB-final-04.fa
├── output
│   └── updated_assemblies
│       ├── Ref.fa
│       ├── LineA.fa
│       └── LineB.fa
└── vcf_dbs
    ├── assemblies.agc *
    ├── gvcf_dataset/
    ├── hvcf_dataset/
    └── temp/
```

> **Note**
>
> In order for this to work, this directory/database **must** contain a
> compressed sequence file. By default, this file will be called `assemblies.agc`.
> This is a compressed file generated by the 
> [Assembled Genomes Compressor](https://github.com/refresh-bio/agc).



### Setup - add AGC path
For **step 2**, we must also have a working AGC application for our operating 
system and environment. This can be established in a couple of ways:


#### Option 1 - Conda
If you went through the steps of setting up a PHG database on your working 
machine, we can _assume_ that you have:

* A working [Conda](https://anaconda.org/anaconda/conda) installation
* A working PHGv2 Conda environment. This by 
  [default](https://phg.maizegenetics.net/build_and_load/#set-up-conda-environment) 
  will be called `phgv2-conda`.

If this is the case, you will, by default, have AGC already set up on your
machine since this is a necessary application for several crucial PHGv2 steps.
You can verify AGC is present in your PHGv2 environment by running the following
command within a terminal:

```
conda run -n phgv2-conda agc
```

If present, this will output a series of commands used by the `agc` program.
For example:

```
$ conda run -n phgv2-conda agc

AGC (Assembled Genomes Compressor) v. 3.1.0 [build 20240312.1]
Usage: agc <command> [options]
Command:
   create   - create archive from FASTA files
   append   - add FASTA files to existing archive
   getcol   - extract all samples from archive
   getset   - extract sample from archive
   getctg   - extract contig from archive
   listref  - list reference sample name in archive
   listset  - list sample names in archive
   listctg  - list sample and contig names in archive
   info     - show some statistics of the compressed data
Note: run agc <command> to see command-specific options
```

> **Note**
>
> If you have changed the name of your environment, the value given with the
> `-n` flag will have to change!

If your machine matches the prior criteria, you may use the rPHG2 function
`linkCondaBin()` to _link_ your Conda environment's AGC application path to
R's global options:

```{r, echo=TRUE, eval=FALSE}
linkCondaBin(condaPath = "/path/to/my/conda/installation")
```

This function will assume your environment name is called `phgv2-conda`. If you
have a different, you can add the modified id to the `envName` parameter.

```{r, echo=TRUE, eval=FALSE}
linkCondaBin(
    condaPath = "/path/to/my/conda/installation",
    envName   = "phgv2-sorghum" # example Conda environment name
)
```


#### Option 2 - manual AGC path
If you do not have Conda installed or no PHGv2 instance, **but** have AGC 
[installed on your machine](https://github.com/refresh-bio/agc/releases), you can manually
set the path using R's `option()` function. In order for rPHG2 to correctly
identify the correct path, you must make a new key in the global options called
`phgv2_agc_path`:

```{r, echo=TRUE, eval=FALSE}
options("phgv2_agc_path" = "/path/to/my/agc/application/bin/agc")
```

> **Note**
>
> This must be the path to the AGC binary (i.e., the file named `agc`) found in 
> the release.



### Setup - build `PHGDataset` object

Now that all prerequisites are completed, we can finish building the 
`PHGDataSet` object using conventional rPHG2 syntax:

```{r, echo=TRUE, eval=FALSE}
phgDs <- localCon
    buildHaplotypeGraph() |>
    readPhgDataSet()
```

```{r, echo=FALSE, eval=FALSE}
phgDs
```



### Using `readSequence()`
Now that we have our `PHGDataSet` object created, we can return specific
sequence information relating to an individual haplotype or a collection of
haplotypes using a haplotype or reference range identifier, respectively.
Similar to other syntax, we can simply pipe our `PHGDataSet` object into the
`readSequence()` method.

For starters, I will return a single haplotype sequence using the `hapId` 
parameter:

```{r, echo=TRUE, eval=TRUE}
# Return first haplotype ID in matrix (example)
hapId <- phgDs |> readHapIds() |> _[1, 1]

phgDs |> readSequence(hapId = hapId)
```

The data object that is returned from `readsequence()` is a `dnastringset`
object which is from the 
[`Biostrings` package](https://bioconductor.org/packages/release/bioc/html/biostrings.html)
found on bioconductor. you can think of this object a list of `dnastringset`
objects which are memory-efficient representations of biological sequences 
(e.g., dna). this will allow users the ability to access all `Biostrings`-related
methods for objects of type `dnastring`. For more information about how to use
these objects, take a look at the 
["Quick Overview" documentation](https://bioconductor.org/packages/release/bioc/vignettes/Biostrings/inst/doc/BiostringsQuickOverview.pdf)
found in the `Biostrings` package.

If we are interested in returning haplotype sequence data from an entire
reference range, we can specify a reference range ID using the `rrId` parameter.
This identifier will be a `character` object containing the following pattern:

```
<chromosome_id>:<start_bp>-<end_bp>
```

For example, this identifier will look like the following "first" reference
range found in our example hVCF data:
```{r, echo=TRUE, eval=TRUE}
# Return first reference range in matrix (example)
rrId <- phgDs |> readHapIds() |> colnames() |> _[1]

rrId
```

```{r, echo=TRUE, eval=TRUE}
phgDs |> readSequence(rrId = rrId)
```


