---
title: "Introduction to rPHG2"
output: rmarkdown::html_vignette
description: >
  Learn how to get started with the basics of rPHG2.
vignette: >
  %\VignetteIndexEntry{Introduction to rPHG2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
editor_options:
  markdown:
    wrap: 70
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.align = "center",
  external = TRUE,
  echo = TRUE,
  warning = FALSE
)

library(rPHG2)
library(GenomicRanges)

source("../tests/testthat/setup.R")

options("phgv2_agc_path" = agcVanillaPath)

hVcfFiles <- system.file(
    "extdata", 
    c("LineA.h.vcf", "LineB.h.vcf"), 
    package = "rPHG2"
)

```

The basic workflow of the `rPHG2` package is as follows:

1.  Create a connection object
2.  Read data into the R environment
3.  Analyze and visualize data retrieval

This document introduces you to `rPHG2`'s methods and grammar, and
shows you how to apply them to the previously mentioned workflow.



## Creating connection objects
rPHG objects can be created through two primary sources:
* local data
* server connections

Creating initial "connection" objects helps unify downstream reading and
evaluation steps for PHGv2 data. In the next couple of sections, we will show
you how to create either local or server connnection objects.


### Local data
_Local_ connections are for local TileDB instances or direct locations of 
[hVCF](https://github.com/maize-genetics/phg_v2/blob/main/docs/hvcf_specifications.md)
files on a local disk. To create a local connection, we will create a
`PHGLocalCon` object using the following constructor function:

```{r, eval=FALSE, echo=TRUE}
PHGLocalCon()
```

For the above example, we will create a connection to some example local hVCF 
files provided with the `rPHG2` package:

* `LineA.h.vcf`
* `LineB.h.vcf`

Since the full paths to these files will differ between each user, we can
use the `system.file()` function to get the full path:

```{r, eval=FALSE, echo=TRUE}
system.file("extdata", "LineA.h.vcf", package = "rPHG2")
```

This will get the full path from the `extdata` directory for the file, 
`LineA.h.vcf`, found in the `rPHG2` source code.

We can further build on this to create a collection of full file paths to our
hVCF data:

```{r, eval=TRUE, echo=TRUE}
hVcfFiles <- system.file(
    "extdata", 
    c("LineA.h.vcf", "LineB.h.vcf"), 
    package = "rPHG2"
)
```


Now that we have a collection of hVCF files, we can use the `PHGLocalCon()`
constructor function to create a `PHGLocalCon` object:

```{r, eval=TRUE, echo=TRUE}
localCon <- hVcfFiles |> PHGLocalCon()

localCon
```

Additionally, if you have a **directory** of valid hVCF files, you can simply
pass the directory path to the constructor. In the following example, I will
point to the `extdata` data directory _without_ pointing to singular hVCF files
that is found within the package:

```{r, eval=TRUE, echo=TRUE}
system.file("extdata", package = "rPHG2") |> PHGLocalCon()
```

From here, we can move to the [next section](#creating-jvm-objects) to create a `HaplotypeGraph`
object interface with the JVM.


### Server connections

> **Note**
>
> We are still actively working on this section and will not have the same
> compatibility as [rPHG](https://rphg.maizegenetics.net/articles/rphg_walkthrough.html#establishing-a-server-connection).
> Stay tuned for further details!

Conversely, _server_ locations are for databases served on publicly available
web services leveraging the [Breeding API (BrAPI) endpoints](https://brapi.org/).
Since this is a connection to a server, a URL path instead of a local file
path will be needed. We will use the `PHGServerCon()` constructor function
to create a `PHGServerCon` object:

```{r, eval=TRUE, echo=TRUE}
srvCon <- "phg.maizegdb.org" |> PHGServerCon()

srvCon
```

In the above example, we have made the assumptions that this URL:

1. Uses a secure transfer protocol ([`https`](https://en.wikipedia.org/wiki/HTTPS))
2. Uses default ports for data serving
3. Has BrAPI specified endpoints

For points **1** and **2**, if the URL uses non-secure protocols ("http")
and/or has a modified port number, you will need to specify these with the
`protocol` and `port` parameters in the constructor function. For example:

```{r, eval=FALSE, echo=TRUE}
"www.my-unsecure-phg.org" |> 
    PHGServerCon(
        port     = 5300, 
        protocol = "http"
    )
```

For point **3**, if the constructor cannot resolve 
[mandatory endpoints](https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core/2.1#/Server%20Info/get_serverinfo),
an exception will occur.



## Creating JVM objects

> **Note**
>
> This will require the `rJava` package to be installed along with a modern
> version of Java (e.g., $\geq$ v17) to work properly!

Now that we have either a local or server-based connection object, we can
convert the raw hVCF data into a 
[`HaplotypeGraph`](https://github.com/maize-genetics/phg_v2/blob/main/src/main/kotlin/net/maizegenetics/phgv2/api/HaplotypeGraph.kt)
JVM object and bridge the Java reference pointer to R. Before we build the JVM 
graph object, we need to initialize the JVM and add the JAR files to our 
environment that are found within the [latest distribution](https://github.com/maize-genetics/phg_v2/releases) 
of PHGv2. 

**Note:** If you have **not** downloaded this, _please_ see instructions
[here](https://phg.maizegenetics.net/installation/#get-phgv2) before you
continue! 

To initialize, run the following command:

```{r, eval=FALSE, echo=TRUE}
initPhg("phg/path/to/lib")
```

...where `phg/path/to/lib` is the `lib` directory found within the decompressed
release of PHGv2. Now that the JVM has been initialized, we can build the JVM
graph using `buildHaplotypeGraph()` using the local connection object as input:


```{r, eval=TRUE, echo=TRUE}
graph <- localCon |> buildHaplotypeGraph()

graph
```

In the above example, we took the local connection object and passed it into
a `HaplotypeGraph` constructor. Here, we have a basic class that contains a
pointer object where we can direct data from Java to R:

```{r, eval=TRUE, echo=TRUE}
graph |> javaRefObj()
```

### Returning version and memory values
Since we need to construct an interface to a local instance of Java, errors
may arise due to several issues. Two common causes are version issues and
memory allocated to the JVM. For debugging and monitoring purposes, we can
use the `jvmStats()` function which creates a instance of a `JvmStats` object.

```{r, eval=TRUE, echo=TRUE}
javaStats <- jvmStats()

javaStats
```

This object contains several values:

* Total number of PHGv2 JAR files added to the class path
* Your local Java version
* The current PHGv2 version added to the class path
* Current memory allocation to the JVM (recorded in gigabytes [`GB`])



## Reading data
Now that we have created a `HaplotypeGraph` object, we can begin **read**ing
data using the `read*` family of `rPHG2` functions.


### Sample IDs
To return a vector of sample IDs from the graph object, we can use the 
`readSamples()` function:

```{r, eval=TRUE, echo=TRUE}
graph |> readSamples()
```


### Reference ranges
To return information about all reference ranges found within the graph object,
we can use the `readRefRanges()` function. This will return a `GRanges` object
which is a common data class in the 
[`GenomicRanges`](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) 
package:

```{r, eval=TRUE, echo=TRUE}
graph |> readRefRanges()
```


### Haplotype IDs
To return all haplotype IDs as a "sample $\times$ reference range" `matrix`
object, we can use the `readHapIds()` function:


```{r, eval=TRUE, echo=TRUE}
m <- graph |> readHapIds()

# Show only first 3 columns
m[, 1:3]
```



### Haplotype ID metadata
To return metadata for each haplotype ID as a `tibble` object, we can use 
the `readHapIdMetaData()` function:

```{r, eval=TRUE, echo=TRUE}
graph |> readHapIdMetaData()
```


### Haplotype ID metadata (positions)
To return positional information for each haplotype ID as another `tibble`
object, we can use the `readHapIdPosMetaData()` function:

```{r, eval=TRUE, echo=TRUE}
graph |> readHapIdPosMetaData()
```


### All hVCF data
In a majority of cases, we may need more than one piece of hVCF data. We can
read all of the prior data simultaneously as `PHGDataSet` object which is
an in-R-memory representation of all hVCF data:

```{r, eval=TRUE, echo=TRUE}
phgDs <- graph |> readPhgDataSet()
phgDs
```

If this object is created, we can use the prior `read*` methods to 
instantaneously pull out the previously mentioned R data objects:

* `readSamples()`
* `readRefRanges()`
* `readHapIds()`
* `readHapIdMetaData()`
* `readHapIdPosMetaData()`



## Filter data
In some cases, you may want to query information and focus on one
specific reference range and/or sample(s). We can **filter** our `PHGDataSet`
objects using the `filter*` family of functions.

For the following examples, let's picture our primary data as a 2-dimensional
`matrix` of:

* samples (rows)
* reference ranges (columns)
* hap ID data (elements)

If we were to represent this as an object in R called `phgDs`, it would look 
like the following diagram:

```{r, eval=TRUE, echo=TRUE}
phgDs
```

![](img/filtering/mat_01.svg){fig.align="center"}

...where each individual colored cell is a haplotype ID.


### Filter by sample
If we want to filter based on sample ID, we can use the `filterSamples()`
function. Simply add the sample ID or collection of sample IDs as a `character`
string or a vector of `character` strings, respectively. Using the prior
diagram as a reference, I will filter out anything that is not the following:

* `B73`
* `Ky21`
* `Mo17`

```{r, eval=FALSE, echo=TRUE}
phgDs |> filterSamples(c("B73", "Ky21", "Mo17"))
```

![](img/filtering/mat_02.svg){fig.align="center"}

**Note**: If samples are added to the filter collection, but are not
present in the data, they will be discarded. If **no** samples are found, an
exception will be thrown stating that no samples could be found.


### Filter by reference range
If we want to filter based on sample ID, we can use the `filterRefRanges()`
function. Currently, this takes a `GRanges` object where we can specify
integer-based ranges located in genomic regions by chromosome (i.e., 
`seqnames`). For example, if I want to return all ranges that intersect with
the following genomic regions:

| Chromosome | Start (bp) | End (bp) |
|------------|------------|----------|
| `"1"`      | `100`      | `400`    |
| `"2"`      | `400`      | `900`    |

I can create the following `GRanges` object and pass that to the filter
method:

```{r, eval=FALSE, echo=TRUE}
gr <- GRanges(
    seqnames = c("1", "2"),
    ranges = IRanges(
        c(100, 800),
        c(400, 900)
    )
)


phgDs |> filterRefRanges(gr)
```

![](img/filtering/mat_03.svg)


### Chaining methods
Since we have two dimensions, we can filter simultaneously by "piping" or
combining filter methods in one pass:

```{r, eval=FALSE, echo=TRUE}
gr <- GRanges(
    seqnames = c("1", "2"),
    ranges = IRanges(
        c(100, 800),
        c(400, 900)
    )
)

phgDs |> 
    filterSamples(c("B73", "Ky21", "Mo17")) |> 
    filterRefRanges(gr)
```

![](img/filtering/mat_04.svg)



## Summarize and visualize data

In addition to parsing multiple hVCF files into R data objects via a 
`PHGDataSet`, rPHG2 also provides functions to summarize and visualize data.


### Global values

To begin, we can return general values on the number of observations within
our dataset using the `numberOf*` family of functions:

```{r, eval=TRUE, echo=TRUE}
phgDs

# Get number of samples/taxa
phgDs |> numberOfSamples()

# Get number of chromosomes
phgDs |> numberOfChromosomes()

# Get number of reference ranges
phgDs |> numberOfRefRanges()

# Get number of haplotype IDs
phgDs |> numberOfHaplotypes()
```


### Get counts of unique haplotype IDs
To get the unique number of haplotypes per reference range in our `PHGDataSet`
object, we can use `numberOfHaplotypes()` again, but set the internal parameter
`byRefRange` to `TRUE`:

```{r, eval=TRUE, echo=TRUE}
phgDs |> numberOfHaplotypes(byRefRange = TRUE)
```


### Visualize counts of unique haplotype IDs
We can visualize the prior section using a member of the `plot*` family of
functions, `plotHaploCounts()`. By default, this will produce a scatter plot of
data found across all reference ranges in the reference genome (similar to a
[Manhattan plot](https://en.wikipedia.org/wiki/Manhattan_plot)):

```{r, eval=TRUE, echo=TRUE}
phgDs |> plotHaploCounts()
```

To get more "granular" views of specific regions within our data, we can
pass a `GRanges` object from the package 
[`GenomicRanges`](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)
to the `gr` parameter:

```{r, eval=TRUE, echo=TRUE}
# library(GenomicRanges)

query <- GRanges(
    seqnames = "1", 
    ranges = IRanges(start = 50, end = 10000)
)

phgDs |> plotHaploCounts(gr = query)
```

We can also query mutltiple regions of interest simultaneously by adding more
observations to the `query` object:

```{r, eval=TRUE, echo=TRUE}
# library(GenomicRanges)

query <- GRanges(
    seqnames = c("1", "2"), 
    ranges = IRanges(start = c(50, 50), end = c(10000, 7500))
)

phgDs |> plotHaploCounts(gr = query)
```

Finally, we provide 3 separate "geometry" options for plotting using the `geom`
parameter:
* `"l"` - line plotting (e.g., `geom_line()`) (_default_)
* `"b"` - bar plotting (e.g., `geom_bar()`)
* `"p"` - point plotting (e.g., `geom_point()`)

For example, if we want to switch this from a line plot to a bar plot, we can
specify `geom = "b"`:

```{r, eval=TRUE, echo=TRUE}
phgDs |> plotHaploCounts(gr = query, geom = "b")
```


### Visualize distributions of unique haplotype IDs
To get a more "global" estimate of unique haplotype ID distribution across the
genome, we can us the function `plotHaploDist()`:

```{r, eval=TRUE, echo=TRUE}
phgDs |> plotHaploDist()
```



## Returning sequence data

We can also return haplotype sequence information via the `readSequence()`
command with PHG data sets (i.e., `PHGDataSet` objects) pointed to [PHGv2 databases
containing compressed assemblies information](https://phg.maizegenetics.net/build_and_load/#compress-fasta-files).

In order to perform sequence retrieval, we must perform a couple of steps:

1. Add `host` information to a `PHGLocalCon` object
2. Add an AGC path to the global options (i.e., the `options()` function)



### Setup - add host information
For **step 1**, during the the construction of a `PHGLocalCon` object using
the `PHGLocalCon()` function, add the path to a 
[PHGv2 database](https://phg.maizegenetics.net/build_and_load/#initialize-tiledb-instances) 
containing a compressed sequence file: 

```{r, echo=TRUE, eval=FALSE}
dbPath <- "/my/phgv2/db_path/"

# using 'hVcfFiles' variable from earlier steps
localCon <- hVcfFiles |> PHGLocalCon(dbUri = dbPath)
```

If you are unsure what this path is, it should be the directory that was
created during the PHGv2 building and loading steps and contains files and
sub-directories that looks like the following:

```
phg_v2_example/
├── data
│   ├── anchors.gff
│   ├── annotation_keyfile.txt
│   ├── Ref-v5.fa
│   ├── LineA-final-01.fa
│   └── LineB-final-04.fa
├── output
│   └── updated_assemblies
│       ├── Ref.fa
│       ├── LineA.fa
│       └── LineB.fa
└── vcf_dbs
    ├── assemblies.agc *
    ├── gvcf_dataset/
    ├── hvcf_dataset/
    └── temp/
```

> **Note**
>
> In order for this to work, this directory/database **must** contain a
> compressed sequence file. By default, this file will be called `assemblies.agc`.
> This is a compressed file generated by the 
> [Assembled Genomes Compressor](https://github.com/refresh-bio/agc).



### Setup - add AGC path
For **step 2**, we must also have a working AGC application for our operating 
system and environment. This can be established in a couple of ways:


#### Option 1 - Conda
If you went through the steps of setting up a PHG database on your working 
machine, we can _assume_ that you have:

* A working [Conda](https://anaconda.org/anaconda/conda) installation
* A working PHGv2 Conda environment. This by 
  [default](https://phg.maizegenetics.net/build_and_load/#set-up-conda-environment) 
  will be called `phgv2-conda`.

If this is the case, you will, by default, have AGC already set up on your
machine since this is a necessary application for several crucial PHGv2 steps.
You can verify AGC is present in your PHGv2 environment by running the following
command within a terminal:

```
conda run -n phgv2-conda agc
```

If present, this will output a series of commands used by the `agc` program.
For example:

```
$ conda run -n phgv2-conda agc

AGC (Assembled Genomes Compressor) v. 3.1.0 [build 20240312.1]
Usage: agc <command> [options]
Command:
   create   - create archive from FASTA files
   append   - add FASTA files to existing archive
   getcol   - extract all samples from archive
   getset   - extract sample from archive
   getctg   - extract contig from archive
   listref  - list reference sample name in archive
   listset  - list sample names in archive
   listctg  - list sample and contig names in archive
   info     - show some statistics of the compressed data
Note: run agc <command> to see command-specific options
```

> **Note**
>
> If you have changed the name of your environment, the value given with the
> `-n` flag will have to change!

If your machine matches the prior criteria, you may use the rPHG2 function
`linkCondaBin()` to _link_ your Conda environment's AGC application path to
R's global options:

```{r, echo=TRUE, eval=FALSE}
linkCondaBin(condaPath = "/path/to/my/conda/installation")
```

This function will assume your environment name is called `phgv2-conda`. If you
have a different, you can add the modified id to the `envName` parameter.

```{r, echo=TRUE, eval=FALSE}
linkCondaBin(
    condaPath = "/path/to/my/conda/installation",
    envName   = "phgv2-sorghum" # example Conda environment name
)
```


#### Option 2 - manual AGC path
If you do not have Conda installed or no PHGv2 instance, **but** have AGC 
[installed on your machine](https://github.com/refresh-bio/agc/releases), you can manually
set the path using R's `option()` function. In order for rPHG2 to correctly
identify the correct path, you must make a new key in the global options called
`phgv2_agc_path`:

```{r, echo=TRUE, eval=FALSE}
options("phgv2_agc_path" = "/path/to/my/agc/application/bin/agc")
```

> **Note**
>
> This must be the path to the AGC binary (i.e., the file named `agc`) found in 
> the release.



### Setup - build `PHGDataset` object

Now that all prerequisites are completed, we can finish building the 
`PHGDataSet` object using conventional rPHG2 syntax:

```{r, echo=FALSE, eval=TRUE}
phgDs <- PHGLocalCon(hVcfFiles, dbUri = phgLibDir) |> 
    buildHaplotypeGraph() |>
    readPhgDataSet()
```

```{r, echo=TRUE, eval=FALSE}
phgDs <- localCon
    buildHaplotypeGraph() |>
    readPhgDataSet()
```

```{r, echo=TRUE, eval=TRUE}
phgDs
```



### Using `readSequence()`
Now that we have our `PHGDataSet` object created, we can return specific
sequence information relating to an individual haplotype or a collection of
haplotypes using a haplotype or reference range identifier, respectively.
Similar to other syntax, we can simply pipe our `PHGDataSet` object into the
`readSequence()` method.

For starters, I will return a single haplotype sequence using the `hapId` 
parameter:

```{r, echo=TRUE, eval=TRUE}
# Return first haplotype ID in matrix (example)
hapId <- phgDs |> readHapIds() |> _[1, 1]

phgDs |> readSequence(hapId = hapId)
```

The data object that is returned from `readsequence()` is a `dnastringset`
object which is from the 
[`Biostrings` package](https://bioconductor.org/packages/release/bioc/html/biostrings.html)
found on bioconductor. you can think of this object a list of `dnastringset`
objects which are memory-efficient representations of biological sequences 
(e.g., dna). this will allow users the ability to access all `Biostrings`-related
methods for objects of type `dnastring`. For more information about how to use
these objects, take a look at the 
["Quick Overview" documentation](https://bioconductor.org/packages/release/bioc/vignettes/Biostrings/inst/doc/BiostringsQuickOverview.pdf)
found in the `Biostrings` package.

If we are interested in returning haplotype sequence data from an entire
reference range, we can specify a reference range ID using the `rrId` parameter.
This identifier will be a `character` object containing the following pattern:

```
<chromosome_id>:<start_bp>-<end_bp>
```

For example, this identifier will look like the following "first" reference
range found in our example hVCF data:
```{r, echo=TRUE, eval=TRUE}
# Return first reference range in matrix (example)
rrId <- phgDs |> readHapIds() |> colnames() |> _[1]

rrId
```

```{r, echo=TRUE, eval=TRUE}
phgDs |> readSequence(rrId = rrId)
```

