<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Load and visualize PHG metrics • rPHG2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Oswald-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Load and visualize PHG metrics">
<meta name="description" content="Learn how to load and visulaize PHG metrics
">
<meta property="og:description" content="Learn how to load and visulaize PHG metrics
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rPHG2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rPHG2.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/metrics.html">Load and visualize PHG metrics</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/rPHG2/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/maize-genetics/phg_v2" aria-label="PHG Source"><span class="fa fa-dna"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Load and visualize PHG metrics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rPHG2/blob/main/vignettes/metrics.Rmd" class="external-link"><code>vignettes/metrics.Rmd</code></a></small>
      <div class="d-none name"><code>metrics.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="creating-phgmetrics-objects">Creating <code>PHGMetrics</code> objects<a class="anchor" aria-label="anchor" href="#creating-phgmetrics-objects"></a>
</h2>
<p>In addition to loading and inspecting hVCF data, <code>rPHG2</code>
provides classes for quick loading and granular visualization of “PHG
metrics” data. Currently, there are two primary types of PHG
metrics:</p>
<ul>
<li>Alignment files from AnchorWave
<ul>
<li>used for collinearity plotting of samples against the reference
genome</li>
</ul>
</li>
<li>gVCF data metrics
<ul>
<li>general statistics of gVCFs generated for each sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>×</mo><annotation encoding="application/x-tex">\times</annotation></semantics></math>
chromosome pair</li>
</ul>
</li>
</ul>
<p>To provide “easy to use” downstream plotting methods, we will need to
construct a new <code>rPHG2</code> <code>PHGMetrics</code> object by
providing either file paths, directory paths, or a combination of the
two as a <code>character</code> vector. In the following example, I will
provide a path to a directory containing both alignment and gVCF data
that is found within the library:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metricDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"rPHG2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metricDir</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/home/runner/work/_temp/Library/rPHG2/extdata"</span></span></code></pre>
<p>From this variable, we can simply pass this to the
<code>PHGMetrics</code> constructor, <code><a href="../reference/PHGMetrics.html">PHGMetrics()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PHGMetrics.html">PHGMetrics</a></span><span class="op">(</span><span class="va">metricDir</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BB00;"> ✔ reading data for: toy_gvcf_metrics.tsv</span></span></span>
<span><span class="co">## <span style="color: #00BB00;"> ✔ reading data for: toy_anchors_s01.anchorspro</span></span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span></span></code></pre></div>
<pre><code><span><span class="co">## A <span style="font-weight: bold;">PHGMetrics</span> object containing <span style="font-weight: bold;">2</span> tables:</span></span>
<span><span class="co">## -- AnchorWave data:</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> toy_anchors_s01 (<span style="color: #808080;">969 10</span>)</span></span>
<span><span class="co">## -- gVCF data:</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> toy_gvcf_metrics (<span style="color: #808080;">41 19</span>)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="accessing-metrics-data-tables">Accessing metrics data tables<a class="anchor" aria-label="anchor" href="#accessing-metrics-data-tables"></a>
</h2>
<div class="section level3">
<h3 id="general-use">General use<a class="anchor" aria-label="anchor" href="#general-use"></a>
</h3>
<p>From the previously created <code>metData</code> object, we can
directly access various types of data through a series of accessor
methods.</p>
<p>If we want to extract all metrics tables as
<code>data.frame</code>-like objects, we can use the
<code><a href="../reference/metricsTable.html">metricsTable()</a></code> function, which will return a
<code>list</code> object of <code>data.frame</code> tables:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsTable.html">metricsTable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $toy_anchors_s01</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 969 × 10</span></span></span>
<span><span class="co">##    ref_chr reference_start reference_end query_chr query_start query_end strand</span></span>
<span><span class="co">##      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>       1           <span style="text-decoration: underline;">52</span>892         <span style="text-decoration: underline;">53</span>561         1       <span style="text-decoration: underline;">22</span>637     <span style="text-decoration: underline;">23</span>277 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>       1          <span style="text-decoration: underline;">275</span>401        <span style="text-decoration: underline;">278</span>445         1      <span style="text-decoration: underline;">196</span>230    <span style="text-decoration: underline;">199</span>275 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>       1          <span style="text-decoration: underline;">665</span>653        <span style="text-decoration: underline;">669</span>559         1      <span style="text-decoration: underline;">456</span>629    <span style="text-decoration: underline;">460</span>535 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>       1          <span style="text-decoration: underline;">723</span>978        <span style="text-decoration: underline;">724</span>529         1      <span style="text-decoration: underline;">514</span>619    <span style="text-decoration: underline;">515</span>170 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>       1         1<span style="text-decoration: underline;">031</span>998       1<span style="text-decoration: underline;">033</span>805         1      <span style="text-decoration: underline;">788</span>012    <span style="text-decoration: underline;">789</span>822 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>       1         1<span style="text-decoration: underline;">497</span>930       1<span style="text-decoration: underline;">503</span>054         1     1<span style="text-decoration: underline;">250</span>572   1<span style="text-decoration: underline;">255</span>668 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>       1         1<span style="text-decoration: underline;">517</span>777       1<span style="text-decoration: underline;">519</span>239         1     1<span style="text-decoration: underline;">270</span>343   1<span style="text-decoration: underline;">271</span>805 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>       1         2<span style="text-decoration: underline;">466</span>395       2<span style="text-decoration: underline;">468</span>560         1     2<span style="text-decoration: underline;">047</span>907   2<span style="text-decoration: underline;">050</span>072 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>       1         2<span style="text-decoration: underline;">533</span>102       2<span style="text-decoration: underline;">536</span>854         1     2<span style="text-decoration: underline;">093</span>924   2<span style="text-decoration: underline;">097</span>674 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>       1         2<span style="text-decoration: underline;">638</span>837       2<span style="text-decoration: underline;">639</span>286         1     2<span style="text-decoration: underline;">188</span>800   2<span style="text-decoration: underline;">189</span>249 +     </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 959 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: gene &lt;chr&gt;, block_index &lt;int&gt;, score &lt;dbl&gt;</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $toy_gvcf_metrics</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 41 × 19</span></span></span>
<span><span class="co">##    taxa  chrom ref_length num_snps num_ins num_del  num_ns num_bases_inserted</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Xb01  ALL    473<span style="text-decoration: underline;">519</span>425  2<span style="text-decoration: underline;">706</span>357  <span style="text-decoration: underline;">336</span>754  <span style="text-decoration: underline;">301</span>867 2<span style="text-decoration: underline;">101</span>168           41<span style="text-decoration: underline;">462</span>213</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Xb01  Vu01    42<span style="text-decoration: underline;">129</span>361   <span style="text-decoration: underline;">154</span>059   <span style="text-decoration: underline;">26</span>280   <span style="text-decoration: underline;">20</span>211  <span style="text-decoration: underline;">181</span>091            2<span style="text-decoration: underline;">600</span>647</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Xb01  Vu02    33<span style="text-decoration: underline;">908</span>088   <span style="text-decoration: underline;">178</span>101   <span style="text-decoration: underline;">21</span>457   <span style="text-decoration: underline;">18</span>981  <span style="text-decoration: underline;">152</span>494            2<span style="text-decoration: underline;">293</span>028</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Xb01  Vu03    65<span style="text-decoration: underline;">292</span>630   <span style="text-decoration: underline;">273</span>800   <span style="text-decoration: underline;">43</span>347   <span style="text-decoration: underline;">38</span>073  <span style="text-decoration: underline;">284</span>407            2<span style="text-decoration: underline;">694</span>624</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Xb01  Vu04    42<span style="text-decoration: underline;">731</span>077   <span style="text-decoration: underline;">424</span>488   <span style="text-decoration: underline;">34</span>561   <span style="text-decoration: underline;">31</span>514  <span style="text-decoration: underline;">177</span>567            3<span style="text-decoration: underline;">908</span>337</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Xb01  Vu05    48<span style="text-decoration: underline;">746</span>289   <span style="text-decoration: underline;">183</span>067   <span style="text-decoration: underline;">28</span>355   <span style="text-decoration: underline;">24</span>918  <span style="text-decoration: underline;">224</span>956            4<span style="text-decoration: underline;">352</span>143</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Xb01  Vu06    34<span style="text-decoration: underline;">463</span>471   <span style="text-decoration: underline;">149</span>413   <span style="text-decoration: underline;">22</span>663   <span style="text-decoration: underline;">20</span>431  <span style="text-decoration: underline;">152</span>644            2<span style="text-decoration: underline;">262</span>451</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Xb01  Vu07    40<span style="text-decoration: underline;">876</span>636   <span style="text-decoration: underline;">164</span>945   <span style="text-decoration: underline;">26</span>583   <span style="text-decoration: underline;">23</span>708  <span style="text-decoration: underline;">178</span>058            1<span style="text-decoration: underline;">966</span>857</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Xb01  Vu08    38<span style="text-decoration: underline;">363</span>498   <span style="text-decoration: underline;">191</span>440   <span style="text-decoration: underline;">26</span>010   <span style="text-decoration: underline;">24</span>047  <span style="text-decoration: underline;">171</span>568            2<span style="text-decoration: underline;">214</span>889</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Xb01  Vu09    43<span style="text-decoration: underline;">933</span>251   <span style="text-decoration: underline;">191</span>892   <span style="text-decoration: underline;">27</span>914   <span style="text-decoration: underline;">23</span>807  <span style="text-decoration: underline;">187</span>467            3<span style="text-decoration: underline;">967</span>889</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 31 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 11 more variables: num_bases_deleted &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   percent_identity_with_ref &lt;dbl&gt;, percent_mapped_to_ref &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mean_insertion_size &lt;dbl&gt;, median_insertion_size &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   largest_insertion &lt;int&gt;, mean_deletion_size &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   median_deletion_size &lt;dbl&gt;, largest_deletion &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ref_ranges_with_haplotype &lt;int&gt;, haplotypes_identical_to_ref &lt;int&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="individual-table-selection">Individual table selection<a class="anchor" aria-label="anchor" href="#individual-table-selection"></a>
</h3>
<p>If we want to extract just one table, we can provide the
<strong>name</strong> of the table into the <code><a href="../reference/metricsTable.html">metricsTable()</a></code>
function by specifying a valid name with the <code>name</code>
parameter. By default, the <code><a href="../reference/PHGMetrics.html">PHGMetrics()</a></code> constructor will
specify the names of each table by extracting the base name of each file
passed to the constructor. For example, if we have the following file
path:</p>
<p><code>/path/to/my/alignment_file.anchorspro</code></p>
<p>…the default name for this table will be <code>alignment_file</code>.
Depending on the complexity of the file ID, <code>rPHG2</code> provides
a <code>$</code> operator override which allows users to cycle through
valid metrics table names:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsTable.html">metricsTable</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">metData</span><span class="op">$</span><span class="va">toy_anchors_s01</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 969 × 10</span></span></span>
<span><span class="co">##    ref_chr reference_start reference_end query_chr query_start query_end strand</span></span>
<span><span class="co">##      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>       1           <span style="text-decoration: underline;">52</span>892         <span style="text-decoration: underline;">53</span>561         1       <span style="text-decoration: underline;">22</span>637     <span style="text-decoration: underline;">23</span>277 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>       1          <span style="text-decoration: underline;">275</span>401        <span style="text-decoration: underline;">278</span>445         1      <span style="text-decoration: underline;">196</span>230    <span style="text-decoration: underline;">199</span>275 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>       1          <span style="text-decoration: underline;">665</span>653        <span style="text-decoration: underline;">669</span>559         1      <span style="text-decoration: underline;">456</span>629    <span style="text-decoration: underline;">460</span>535 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>       1          <span style="text-decoration: underline;">723</span>978        <span style="text-decoration: underline;">724</span>529         1      <span style="text-decoration: underline;">514</span>619    <span style="text-decoration: underline;">515</span>170 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>       1         1<span style="text-decoration: underline;">031</span>998       1<span style="text-decoration: underline;">033</span>805         1      <span style="text-decoration: underline;">788</span>012    <span style="text-decoration: underline;">789</span>822 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>       1         1<span style="text-decoration: underline;">497</span>930       1<span style="text-decoration: underline;">503</span>054         1     1<span style="text-decoration: underline;">250</span>572   1<span style="text-decoration: underline;">255</span>668 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>       1         1<span style="text-decoration: underline;">517</span>777       1<span style="text-decoration: underline;">519</span>239         1     1<span style="text-decoration: underline;">270</span>343   1<span style="text-decoration: underline;">271</span>805 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>       1         2<span style="text-decoration: underline;">466</span>395       2<span style="text-decoration: underline;">468</span>560         1     2<span style="text-decoration: underline;">047</span>907   2<span style="text-decoration: underline;">050</span>072 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>       1         2<span style="text-decoration: underline;">533</span>102       2<span style="text-decoration: underline;">536</span>854         1     2<span style="text-decoration: underline;">093</span>924   2<span style="text-decoration: underline;">097</span>674 +     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>       1         2<span style="text-decoration: underline;">638</span>837       2<span style="text-decoration: underline;">639</span>286         1     2<span style="text-decoration: underline;">188</span>800   2<span style="text-decoration: underline;">189</span>249 +     </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 959 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: gene &lt;chr&gt;, block_index &lt;int&gt;, score &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="table-selection-by-metric-type">Table selection by metric type<a class="anchor" aria-label="anchor" href="#table-selection-by-metric-type"></a>
</h3>
<p>We can also extract tables by type using the <code>type</code>
parameter in the <code><a href="../reference/metricsTable.html">metricsTable()</a></code> function. Currently, the
two metric types that PHGv2 provides are as follows:</p>
<ul>
<li>
<code>align</code> - For AnchorWave alignment files</li>
<li>
<code>gvcf</code> - For gVCF metrics files</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsTable.html">metricsTable</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"gvcf"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 41 × 19</span></span></span>
<span><span class="co">##    taxa  chrom ref_length num_snps num_ins num_del  num_ns num_bases_inserted</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Xb01  ALL    473<span style="text-decoration: underline;">519</span>425  2<span style="text-decoration: underline;">706</span>357  <span style="text-decoration: underline;">336</span>754  <span style="text-decoration: underline;">301</span>867 2<span style="text-decoration: underline;">101</span>168           41<span style="text-decoration: underline;">462</span>213</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Xb01  Vu01    42<span style="text-decoration: underline;">129</span>361   <span style="text-decoration: underline;">154</span>059   <span style="text-decoration: underline;">26</span>280   <span style="text-decoration: underline;">20</span>211  <span style="text-decoration: underline;">181</span>091            2<span style="text-decoration: underline;">600</span>647</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Xb01  Vu02    33<span style="text-decoration: underline;">908</span>088   <span style="text-decoration: underline;">178</span>101   <span style="text-decoration: underline;">21</span>457   <span style="text-decoration: underline;">18</span>981  <span style="text-decoration: underline;">152</span>494            2<span style="text-decoration: underline;">293</span>028</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Xb01  Vu03    65<span style="text-decoration: underline;">292</span>630   <span style="text-decoration: underline;">273</span>800   <span style="text-decoration: underline;">43</span>347   <span style="text-decoration: underline;">38</span>073  <span style="text-decoration: underline;">284</span>407            2<span style="text-decoration: underline;">694</span>624</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Xb01  Vu04    42<span style="text-decoration: underline;">731</span>077   <span style="text-decoration: underline;">424</span>488   <span style="text-decoration: underline;">34</span>561   <span style="text-decoration: underline;">31</span>514  <span style="text-decoration: underline;">177</span>567            3<span style="text-decoration: underline;">908</span>337</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Xb01  Vu05    48<span style="text-decoration: underline;">746</span>289   <span style="text-decoration: underline;">183</span>067   <span style="text-decoration: underline;">28</span>355   <span style="text-decoration: underline;">24</span>918  <span style="text-decoration: underline;">224</span>956            4<span style="text-decoration: underline;">352</span>143</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Xb01  Vu06    34<span style="text-decoration: underline;">463</span>471   <span style="text-decoration: underline;">149</span>413   <span style="text-decoration: underline;">22</span>663   <span style="text-decoration: underline;">20</span>431  <span style="text-decoration: underline;">152</span>644            2<span style="text-decoration: underline;">262</span>451</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Xb01  Vu07    40<span style="text-decoration: underline;">876</span>636   <span style="text-decoration: underline;">164</span>945   <span style="text-decoration: underline;">26</span>583   <span style="text-decoration: underline;">23</span>708  <span style="text-decoration: underline;">178</span>058            1<span style="text-decoration: underline;">966</span>857</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Xb01  Vu08    38<span style="text-decoration: underline;">363</span>498   <span style="text-decoration: underline;">191</span>440   <span style="text-decoration: underline;">26</span>010   <span style="text-decoration: underline;">24</span>047  <span style="text-decoration: underline;">171</span>568            2<span style="text-decoration: underline;">214</span>889</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Xb01  Vu09    43<span style="text-decoration: underline;">933</span>251   <span style="text-decoration: underline;">191</span>892   <span style="text-decoration: underline;">27</span>914   <span style="text-decoration: underline;">23</span>807  <span style="text-decoration: underline;">187</span>467            3<span style="text-decoration: underline;">967</span>889</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 31 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 11 more variables: num_bases_deleted &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   percent_identity_with_ref &lt;dbl&gt;, percent_mapped_to_ref &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mean_insertion_size &lt;dbl&gt;, median_insertion_size &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   largest_insertion &lt;int&gt;, mean_deletion_size &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   median_deletion_size &lt;dbl&gt;, largest_deletion &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ref_ranges_with_haplotype &lt;int&gt;, haplotypes_identical_to_ref &lt;int&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="other-accessor-functions">Other accessor functions<a class="anchor" aria-label="anchor" href="#other-accessor-functions"></a>
</h2>
<p>In addition to extracting tables, we can also access other types of
data found in <code>PHGMetrics</code> objects.</p>
<div class="section level3">
<h3 id="identifiers">Identifiers<a class="anchor" aria-label="anchor" href="#identifiers"></a>
</h3>
<p>If we want to return all IDs of every table we can use the
<code><a href="../reference/metricsIds.html">metricsIds()</a></code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsIds.html">metricsIds</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "toy_anchors_s01"  "toy_gvcf_metrics"</span></span></code></pre>
<p>Similar to the previous section, we can also return IDs by metric
type using the <code>type</code> parameter:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsIds.html">metricsIds</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"align"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "toy_anchors_s01"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>We can also extract metadata which will provide a simple
<code>data.frame</code> table containing three columns:</p>
<ul>
<li>
<code>file</code> - file ID of loaded data</li>
<li>
<code>type</code> - metrics data type</li>
<li>
<code>id</code> - the “name” of the metrics table</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsMetaData.html">metricsMetaData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">##   file                       type  id              </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> toy_gvcf_metrics.tsv       gvcf  toy_gvcf_metrics</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> toy_anchors_s01.anchorspro align toy_anchors_s01</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="updating-phgmetrics-objects">Updating <code>PHGMetrics</code> objects<a class="anchor" aria-label="anchor" href="#updating-phgmetrics-objects"></a>
</h2>
<p>To update existing <code>PHGMetrics</code> we can use
<code>f()&lt;-</code> complements to the accessor methods from the prior
sections. Currently, we can update or add the following information:</p>
<ul>
<li>metrics table names/IDs (<em>update</em>)</li>
<li>metrics tables (<em>add</em>)</li>
</ul>
<div class="section level3">
<h3 id="update-metric-ids">Update metric IDs<a class="anchor" aria-label="anchor" href="#update-metric-ids"></a>
</h3>
<p>To update pre-existing names of tables in the object, we can use the
<code><a href="../reference/metricsIds.html">metricsIds()</a></code> function in conjunction with the assignment
(<code>&lt;-</code>) operator by passing a named <code>character</code>
vector, where the name of the vector is the name of the old table ID and
the element is the new name:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nameMap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"old_id1"</span> <span class="op">=</span> <span class="st">"new_id1"</span>,</span>
<span>    <span class="st">"old_id2"</span> <span class="op">=</span> <span class="st">"new_id2"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If we do not wish to remember all old IDs of our tables, we can use
the prior <code>$</code> operator in conjunction with
<code>rPHG2</code>’s defined “to” operator, <code>%T%</code> in a
vector:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nameMap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="va">metData</span><span class="op">$</span><span class="va">toy_gvcf_metrics</span> <span class="op"><a href="../reference/grapes-T-grapes.html">%T%</a></span> <span class="st">"new_gvcf_id"</span>,</span>
<span>    <span class="va">metData</span><span class="op">$</span><span class="va">toy_anchors_s01</span> <span class="op"><a href="../reference/grapes-T-grapes.html">%T%</a></span> <span class="st">"new_anchor_id"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The prior code will internally evaluate to named
<code>character</code> vector:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nameMap</span></span></code></pre></div>
<pre><code><span><span class="co">## toy_gvcf_metrics  toy_anchors_s01 </span></span>
<span><span class="co">##    "new_gvcf_id"  "new_anchor_id"</span></span></code></pre>
<p>This can be passed to the <code><a href="../reference/metricsIds.html">metricsIds()</a></code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metricsIds.html">metricsIds</a></span><span class="op">(</span><span class="va">metData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">nameMap</span></span>
<span></span>
<span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsIds.html">metricsIds</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "new_anchor_id" "new_gvcf_id"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="update-chromosome-ids">Update chromosome IDs<a class="anchor" aria-label="anchor" href="#update-chromosome-ids"></a>
</h3>
<p>To update pre-existing chromosome IDs found within a
<code>PHGMetrics</code> object, we can use the <code>seqnames()</code>
in conjunction with the assignment (<code>&lt;-</code>) operator by
passing a <code>data.frame</code> object that contains two columns:</p>
<ul>
<li>
<code>old_id</code> - old chromosome IDs already present in the
object</li>
<li>
<code>new_id</code> - the new name for the ID</li>
</ul>
<p>We can also use this function to return all global chromosome IDs
from the object:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make copy for example purposes</span></span>
<span><span class="va">metNew</span> <span class="op">&lt;-</span> <span class="va">metData</span></span>
<span></span>
<span><span class="va">metNew</span> <span class="op">|&gt;</span> <span class="fu">seqnames</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "1"          "2"          "3"          "Vu01"       "Vu02"      </span></span>
<span><span class="co">##  [6] "Vu03"       "Vu04"       "Vu05"       "Vu06"       "Vu07"      </span></span>
<span><span class="co">## [11] "Vu08"       "Vu09"       "Vu10"       "Vu11"       "contig_206"</span></span>
<span><span class="co">## [16] "contig_178"</span></span></code></pre>
<p>To update, simply pass the <code>data.frame</code> object with the
old and new ID maps to the <code>seqnames()</code> function:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keyFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    old_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vu01"</span>, <span class="st">"Vu03"</span><span class="op">)</span>,</span>
<span>    new_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR_01"</span>, <span class="st">"CHR_03"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">seqnames</span><span class="op">(</span><span class="va">metNew</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">keyFile</span></span>
<span></span>
<span><span class="va">metNew</span> <span class="op">|&gt;</span> <span class="fu">seqnames</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "1"          "2"          "3"          "CHR_01"     "Vu02"      </span></span>
<span><span class="co">##  [6] "CHR_03"     "Vu04"       "Vu05"       "Vu06"       "Vu07"      </span></span>
<span><span class="co">## [11] "Vu08"       "Vu09"       "Vu10"       "Vu11"       "contig_206"</span></span>
<span><span class="co">## [16] "contig_178"</span></span></code></pre>
<p>As you can see, all instances of <code>"Vu01"</code> and
<code>"Vu03"</code> have been globally changed to <code>"CHR_01"</code>
and <code>"CHR_03"</code>.</p>
</div>
<div class="section level3">
<h3 id="update-tables">Update tables<a class="anchor" aria-label="anchor" href="#update-tables"></a>
</h3>
<p>We can also add new tables to our existing <code>PHGMetrics</code>
objects using the <code><a href="../reference/metricsTable.html">metricsTable()</a></code> function by passing a
named <code>list</code> object containing a combination of the valid
types:</p>
<ul>
<li>file paths</li>
<li>directory paths</li>
<li>in-memory R <code>data.frame</code> objects containing valid PHGv2
metric structures</li>
</ul>
<p>For this example, I will add another table via a
<code>data.frame</code> using some internal toy data found in the
<code>rPHG2</code> package:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">algnPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/toy_anchors_s01.anchorspro"</span>, package <span class="op">=</span> <span class="st">"rPHG2"</span><span class="op">)</span></span>
<span><span class="va">algnDf</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">algnPath</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">algnDf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   refChr referenceStart referenceEnd queryChr queryStart queryEnd strand</span></span>
<span><span class="co">## 1      1          52892        53561        1      22637    23277      +</span></span>
<span><span class="co">## 2      1         275401       278445        1     196230   199275      +</span></span>
<span><span class="co">## 3      1         665653       669559        1     456629   460535      +</span></span>
<span><span class="co">## 4      1         723978       724529        1     514619   515170      +</span></span>
<span><span class="co">## 5      1        1031998      1033805        1     788012   789822      +</span></span>
<span><span class="co">## 6      1        1497930      1503054        1    1250572  1255668      +</span></span>
<span><span class="co">##          gene blockIndex   score</span></span>
<span><span class="co">## 1    OQU90574         40 0.83871</span></span>
<span><span class="co">## 2 interanchor         40      NA</span></span>
<span><span class="co">## 3    KXG37092         40 1.00000</span></span>
<span><span class="co">## 4    KXG37098         40 1.00000</span></span>
<span><span class="co">## 5    EER93094         40 1.00000</span></span>
<span><span class="co">## 6    KXG37167         40 1.00000</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/metricsTable.html">metricsTable</a></span><span class="op">(</span><span class="va">metData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"new_algn_table"</span> <span class="op">=</span> <span class="va">algnDf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BB00;"> ✔ importing data for: new_algn_table</span></span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/metricsMetaData.html">metricsMetaData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">##   file                       type  id            </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> toy_gvcf_metrics.tsv       gvcf  new_gvcf_id   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> toy_anchors_s01.anchorspro align new_anchor_id </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> new_algn_table             align new_algn_table</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualize-metrics-data">Visualize metrics data<a class="anchor" aria-label="anchor" href="#visualize-metrics-data"></a>
</h2>
<p>If you are satisfied with how the metrics data is loaded into R, we
can directly visualize granular diagnostic plots using the
<code>plot*</code> family of functions.</p>
<div class="section level3">
<h3 id="anchorwave-data">AnchorWave data<a class="anchor" aria-label="anchor" href="#anchorwave-data"></a>
</h3>
<p>For example, if I want to generate dot plots from alignment data, I
can simply pass the <code>PHGMetrics</code> object to the function,
<code><a href="../reference/plotDot.html">plotDot()</a></code>. Currently this can only plot one table at a
time, so we must specify the alignment table that we want to
visualize:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotDot.html">plotDot</a></span><span class="op">(</span><span class="va">metData</span><span class="op">$</span><span class="va">new_anchor_id</span><span class="op">)</span></span></code></pre></div>
<p><img src="metrics_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<p><strong>Note</strong>: This plot can only work on alignment
(<code>type = "align"</code>) data and will throw an error if any
gVCF-based metrics Ids are used.</p>
<p>If only one AnchorWave table is loaded into memory, a user can also
specify a more simpler approach:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotDot.html">plotDot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note</strong>: If more than one table is loaded, this
approach will still work, but will pick the first table/elemement in the
internal <code>list</code> object.</p>
</div>
<div class="section level3">
<h3 id="gvcf-metrics-data">gVCF metrics data<a class="anchor" aria-label="anchor" href="#gvcf-metrics-data"></a>
</h3>
<p>We can also visualize gVCF data using the function
<code><a href="../reference/plotGvcf.html">plotGvcf()</a></code>. This method provides granular means of
visualizing various forms of metrics data found within the <a href="https://github.com/maize-genetics/phg_v2/blob/main/docs/qc_metrics.md#output" class="external-link">gVCF
metrics file</a>.</p>
<p>This method takes 3 primary components:</p>
<ul>
<li>The <code>PHGMetrics</code> object</li>
<li>A metric ID</li>
<li>A gVCF metric ID
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>∼</mo><annotation encoding="application/x-tex">\sim</annotation></semantics></math>
contig ID formula</li>
</ul>
<p>A simple approach would be to pass the <code>PHGMetrics</code> object
along with the specified metrics ID:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotGvcf.html">plotGvcf</a></span><span class="op">(</span><span class="va">metData</span><span class="op">$</span><span class="va">new_gvcf_id</span><span class="op">)</span></span></code></pre></div>
<p><img src="metrics_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Similar to the alignment approach, if you have just one table, you
can also pass just the function, since it will dynamically pick the only
table:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotGvcf.html">plotGvcf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="formula-usage">Formula usage<a class="anchor" aria-label="anchor" href="#formula-usage"></a>
</h4>
<p>For more “granular” approaches, we can specify a formula
(<code>f</code>) to dictate what types of metrics and chromosome IDs we
want plotted using the following logic:</p>
<pre><code>&lt;gvcf metric ids&gt; ~ &lt;contig ids&gt;</code></pre>
<p>The following are valid metric IDs:</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="74%">
</colgroup>
<thead><tr class="header">
<th>Metric ID</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ref_length</code></td>
<td>The length of the reference sequence</td>
</tr>
<tr class="even">
<td><code>num_snps</code></td>
<td>The number of SNP records in the gVCF file for the given
chromosome</td>
</tr>
<tr class="odd">
<td><code>num_ins</code></td>
<td>The number of insertion records in the gVCF file</td>
</tr>
<tr class="even">
<td><code>num_del</code></td>
<td>The number of deletion records in the gVCF file</td>
</tr>
<tr class="odd">
<td><code>num_ns</code></td>
<td>The number of N’s/ambiguous bases in the assembly alignment</td>
</tr>
<tr class="even">
<td><code>num_bases_inserted</code></td>
<td>The number of bases inserted relative to the reference sequence</td>
</tr>
<tr class="odd">
<td><code>num_bases_deleted</code></td>
<td>The number of bases deleted relative to the reference sequence</td>
</tr>
<tr class="even">
<td><code>percent_identity_with_ref</code></td>
<td>The proportion of bases relative to refLength that are the same base
as the reference base</td>
</tr>
<tr class="odd">
<td><code>percent_mapped_to_ref</code></td>
<td>The proportion of bases relative to refLength that are present in a
gVCF record</td>
</tr>
<tr class="even">
<td><code>mean_insertion_size</code></td>
<td>The mean size of insertion records</td>
</tr>
<tr class="odd">
<td><code>median_insertion_size</code></td>
<td>The median size of insertion records</td>
</tr>
<tr class="even">
<td><code>largest_insertion</code></td>
<td>The size of the largest insertion</td>
</tr>
<tr class="odd">
<td><code>mean_deletion_size</code></td>
<td>The mean size of deletion records</td>
</tr>
<tr class="even">
<td><code>median_deletion_size</code></td>
<td>The median size of deletion records</td>
</tr>
<tr class="odd">
<td><code>largest_deletion</code></td>
<td>The size of the largest deletion</td>
</tr>
</tbody>
</table>
<p>Due to the number of metric IDs, several “collection” IDs are also
valid:</p>
<table class="table">
<colgroup>
<col width="9%">
<col width="90%">
</colgroup>
<thead><tr class="header">
<th>Collection ID</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ALL</code></td>
<td>Plots <strong>all</strong> prior metric IDs</td>
</tr>
<tr class="even">
<td><code>CORE</code></td>
<td>Plots <code>num_snps</code>, <code>num_ins</code>,
<code>num_del</code>, <code>num_ns</code>,
<code>percent_identity_with_ref</code>,
<code>percent_mapped_to_ref</code>, <code>largest_insertion</code>,
<code>largest_deletion</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Note</strong>: Collection IDs may only be plotted one at a
time or else an error will be thrown.</p>
<p>For example, if I want to plot:</p>
<ul>
<li>the length of the reference sequence (<code>ref_length</code>)</li>
<li>the number of deletions (<code>num_del</code>)</li>
</ul>
<p>…for chromsomes:</p>
<ul>
<li><code>Vu01</code></li>
<li><code>Vu05</code></li>
</ul>
<p>…I can use the following formula syntax:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotGvcf.html">plotGvcf</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">ref_length</span> <span class="op">+</span> <span class="va">num_del</span> <span class="op">~</span> <span class="va">Vu01</span> <span class="op">+</span> <span class="va">Vu05</span><span class="op">)</span></span></code></pre></div>
<p><img src="metrics_files/figure-html/unnamed-chunk-21-1.png" width="700" style="display: block; margin: auto;"></p>
<p>I can also plot the cumulative <code>ALL</code> category which
combines all chromosome metrics into one using the <code>ALL</code>
keyword on the right-hand-side of the equation:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotGvcf.html">plotGvcf</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">ref_length</span> <span class="op">+</span> <span class="va">num_del</span> <span class="op">~</span> <span class="va">ALL</span><span class="op">)</span></span></code></pre></div>
<p><img src="metrics_files/figure-html/unnamed-chunk-22-1.png" width="700" style="display: block; margin: auto;"></p>
<p>One thing you may have noticed is that the formula parameter is
<strong>not required</strong>. If no formula is used, the default
equation will be:</p>
<pre><code><span><span class="va">CORE</span> <span class="op">~</span> <span class="va">ALL</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="dimensions">Dimensions<a class="anchor" aria-label="anchor" href="#dimensions"></a>
</h4>
<p>We can also specify the dimensions of the plot components using two
parameters:</p>
<ul>
<li>number of rows (<code>nRow</code>)</li>
<li>number of columns (<code>nCol</code>)</li>
</ul>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotGvcf.html">plotGvcf</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">ref_length</span> <span class="op">+</span> <span class="va">num_del</span> <span class="op">~</span> <span class="va">ALL</span>, nCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="metrics_files/figure-html/unnamed-chunk-23-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="plot-tags">Plot “tags”<a class="anchor" aria-label="anchor" href="#plot-tags"></a>
</h4>
<p>We can also modify the tagging type using symbolic identifiers via
the <code>tag</code> parameter:</p>
<table class="table">
<thead><tr class="header">
<th>Symbol</th>
<th>Sub plot tag progression</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>"A"</code></td>
<td>
<code>"A"</code>, <code>"B"</code>, <code>"C"</code>, …</td>
</tr>
<tr class="even">
<td><code>"a"</code></td>
<td>
<code>"a"</code>, <code>"b"</code>, <code>"c"</code>, …</td>
</tr>
<tr class="odd">
<td><code>"1"</code></td>
<td>
<code>"1"</code>, <code>"2"</code>, <code>"3"</code>, …</td>
</tr>
<tr class="even">
<td><code>"I"</code></td>
<td>
<code>"I"</code>, <code>"II"</code>, <code>"III"</code>, …</td>
</tr>
<tr class="odd">
<td><code>"i"</code></td>
<td>
<code>"i"</code>, <code>"ii"</code>, <code>"iii"</code>, …</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metData</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plotGvcf.html">plotGvcf</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">CORE</span> <span class="op">~</span> <span class="va">ALL</span>, tag <span class="op">=</span> <span class="st">"i"</span><span class="op">)</span></span></code></pre></div>
<p><img src="metrics_files/figure-html/unnamed-chunk-24-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p><b>rPHG2</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
