<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Learn how to get started with the basics of rPHG2.
">
<title>Introduction to rPHG2 • rPHG2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Barlow-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Oswald-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction to rPHG2">
<meta property="og:description" content="Learn how to get started with the basics of rPHG2.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rPHG2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/rPHG2.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/maize-genetics/rPHG2/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/maize-genetics/phg_v2" aria-label="PHG Source">
    <span class="fa fa-dna"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to rPHG2</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/maize-genetics/rPHG2/blob/HEAD/vignettes/rPHG2.Rmd" class="external-link"><code>vignettes/rPHG2.Rmd</code></a></small>
      <div class="d-none name"><code>rPHG2.Rmd</code></div>
    </div>

    
    
<p>The basic workflow of the <code>rPHG2</code> package is as
follows:</p>
<ol style="list-style-type: decimal">
<li>Create a connection object</li>
<li>Read data into the R environment</li>
<li>Analyze and visualize data retrieval</li>
</ol>
<p>This document introduces you to <code>rPHG2</code>’s methods and
grammar, and shows you how to apply them to the previously mentioned
workflow.</p>
<div class="section level2">
<h2 id="creating-connection-objects">Creating connection objects<a class="anchor" aria-label="anchor" href="#creating-connection-objects"></a>
</h2>
<p>rPHG objects can be created through two primary sources: * local data
* server connections</p>
<p>Creating initial “connection” objects helps unify downstream reading
and evaluation steps for PHGv2 data. In the next couple of sections, we
will show you how to create either local or server connnection
objects.</p>
<div class="section level3">
<h3 id="local-data">Local data<a class="anchor" aria-label="anchor" href="#local-data"></a>
</h3>
<p><em>Local</em> connections are for local TileDB instances or direct
locations of <a href="https://github.com/maize-genetics/phg_v2/blob/main/docs/hvcf_specifications.md" class="external-link">hVCF</a>
files on a local disk. To create a local connection, we will create a
<code>PHGLocalCon</code> object using the following constructor
function:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PHGLocalCon.html">PHGLocalCon</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>For the above example, we will create a connection to some example
local hVCF files provided with the <code>rPHG2</code> package:</p>
<ul>
<li><code>LineA.h.vcf</code></li>
<li><code>LineB.h.vcf</code></li>
</ul>
<p>Since the full paths to these files will differ between each user, we
can use the <code><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file()</a></code> function to get the full
path:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"LineA.h.vcf"</span>, package <span class="op">=</span> <span class="st">"rPHG2"</span><span class="op">)</span></span></code></pre></div>
<p>This will get the full path from the <code>extdata</code> directory
for the file, <code>LineA.h.vcf</code>, found in the <code>rPHG2</code>
source code.</p>
<p>We can further build on this to create a collection of full file
paths to our hVCF data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hVcfFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LineA.h.vcf"</span>, <span class="st">"LineB.h.vcf"</span><span class="op">)</span>, </span>
<span>    package <span class="op">=</span> <span class="st">"rPHG2"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now that we have a collection of hVCF files, we can use the
<code><a href="../reference/PHGLocalCon.html">PHGLocalCon()</a></code> constructor function to create a
<code>PHGLocalCon</code> object:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">localCon</span> <span class="op">&lt;-</span> <span class="va">hVcfFiles</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/PHGLocalCon.html">PHGLocalCon</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">localCon</span></span></code></pre></div>
<pre><code><span><span class="co">## A <span style="font-weight: bold;">PHGLocalCon</span> connection object</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> DB URI....   : <span style="color: #555555;">◯</span></span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> hVCF Files...: <span style="color: #00BB00;">◉</span></span></span></code></pre>
<p>From here, we can move to the <a href="#creating-jvm-objects">next
section</a> to create a <code>HaplotypeGraph</code> object interface
with the JVM.</p>
</div>
<div class="section level3">
<h3 id="server-connections">Server connections<a class="anchor" aria-label="anchor" href="#server-connections"></a>
</h3>
<p>Conversely, <em>server</em> locations are for databases served on
publicly available web services leveraging the <a href="https://brapi.org/" class="external-link">Breeding API (BrAPI) endpoints</a>. Since this
is a connection to a server, a URL path instead of a local file path
will be needed. We will use the <code><a href="../reference/PHGServerCon.html">PHGServerCon()</a></code> constructor
function to create a <code>PHGServerCon</code> object:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">srvCon</span> <span class="op">&lt;-</span> <span class="st">"phg.maizegdb.org"</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/PHGServerCon.html">PHGServerCon</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">srvCon</span></span></code></pre></div>
<pre><code><span><span class="co">## A <span style="font-weight: bold;">PHGServerCon</span> connection object</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> Host............: phg.maizegdb.org</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> Server Status...: 200 (<span style="color: #00BB00; font-weight: bold;">OK</span>)</span></span></code></pre>
<p>In the above example, we have made the assumptions that this URL:</p>
<ol style="list-style-type: decimal">
<li>Uses a secure transfer protocol (<a href="https://en.wikipedia.org/wiki/HTTPS" class="external-link"><code>https</code></a>)</li>
<li>Uses default ports for data serving</li>
<li>Has BrAPI specified endpoints</li>
</ol>
<p>For points <strong>1</strong> and <strong>2</strong>, if the URL uses
non-secure protocols (“http”) and/or has a modified port number, you
will need to specify these with the <code>protocol</code> and
<code>port</code> parameters in the constructor function. For
example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="st">"www.my-unsecure-phg.org"</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/PHGServerCon.html">PHGServerCon</a></span><span class="op">(</span></span>
<span>        port     <span class="op">=</span> <span class="fl">5300</span>, </span>
<span>        protocol <span class="op">=</span> <span class="st">"http"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>For point <strong>3</strong>, if the constructor cannot resolve <a href="https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core/2.1#/Server%20Info/get_serverinfo" class="external-link">mandatory
endpoints</a>, an exception will occur.</p>
</div>
</div>
<div class="section level2">
<h2 id="creating-jvm-objects">Creating JVM objects<a class="anchor" aria-label="anchor" href="#creating-jvm-objects"></a>
</h2>
<p>Now that we have either a local or server-based connection object, we
can convert the raw hVCF data into a <a href="https://github.com/maize-genetics/phg_v2/blob/main/src/main/kotlin/net/maizegenetics/phgv2/api/HaplotypeGraph.kt" class="external-link"><code>HaplotypeGraph</code></a>
JVM object and bridge the Java reference pointer to R using the
function, <code><a href="../reference/buildHaplotypeGraph.html">buildHaplotypeGraph()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="va">localCon</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/buildHaplotypeGraph.html">buildHaplotypeGraph</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">graph</span></span></code></pre></div>
<pre><code><span><span class="co">## A <span style="font-weight: bold;">HaplotypeGraph</span> object @ <span style="color: #0000BB; font-weight: bold;">17f9d882</span></span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of ref ranges....: 38</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of taxa..........: 2</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of chromosomes...: 2</span></span></code></pre>
<p>In the above example, we took the local connection object and passed
it into a <code>HaplotypeGraph</code> constructor. Here, we have a basic
class that contains a pointer object where we can direct data from Java
to R:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/javaRefObj.html">javaRefObj</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Java-Object{net.maizegenetics.phgv2.api.HaplotypeGraph@17f9d882}"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="reading-data">Reading data<a class="anchor" aria-label="anchor" href="#reading-data"></a>
</h2>
<p>Now that we have created a <code>HaplotypeGraph</code> object, we can
begin <strong>read</strong>ing data using the <code>read*</code> family
of <code>rPHG2</code> functions.</p>
<div class="section level3">
<h3 id="sample-ids">Sample IDs<a class="anchor" aria-label="anchor" href="#sample-ids"></a>
</h3>
<p>To return a vector of sample IDs from the graph object, we can use
the <code><a href="../reference/readSamples.html">readSamples()</a></code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/readSamples.html">readSamples</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "LineA" "LineB"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="reference-ranges">Reference ranges<a class="anchor" aria-label="anchor" href="#reference-ranges"></a>
</h3>
<p>To return information about all reference ranges found within the
graph object, we can use the <code><a href="../reference/readRefRanges.html">readRefRanges()</a></code> function. This
will return a <code>GRanges</code> object which is a common data class
in the <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link"><code>GenomicRanges</code></a>
package:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/readRefRanges.html">readRefRanges</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 38 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames      ranges strand |         rr_id</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;</span></span>
<span><span class="co">##    [1]        1      1-1000      * |      1:1-1000</span></span>
<span><span class="co">##    [2]        1   1001-5500      * |   1:1001-5500</span></span>
<span><span class="co">##    [3]        1   5501-6500      * |   1:5501-6500</span></span>
<span><span class="co">##    [4]        1  6501-11000      * |  1:6501-11000</span></span>
<span><span class="co">##    [5]        1 11001-12000      * | 1:11001-12000</span></span>
<span><span class="co">##    ...      ...         ...    ... .           ...</span></span>
<span><span class="co">##   [34]        2 38501-39500      * | 2:38501-39500</span></span>
<span><span class="co">##   [35]        2 39501-44000      * | 2:39501-44000</span></span>
<span><span class="co">##   [36]        2 44001-45000      * | 2:44001-45000</span></span>
<span><span class="co">##   [37]        2 45001-49500      * | 2:45001-49500</span></span>
<span><span class="co">##   [38]        2 49501-50500      * | 2:49501-50500</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="haplotype-ids">Haplotype IDs<a class="anchor" aria-label="anchor" href="#haplotype-ids"></a>
</h3>
<p>To return all haplotype IDs as a “sample <span class="math inline">\(\times\)</span> reference range”
<code>matrix</code> object, we can use the <code><a href="../reference/readHapIds.html">readHapIds()</a></code>
function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/readHapIds.html">readHapIds</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Show only first 3 columns</span></span>
<span><span class="va">m</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          1:1-1000                           1:1001-5500                       </span></span>
<span><span class="co">## LineA_G1 "12f0cec9102e84a161866e37072443b7" "3149b3144f93134eb29661bade697fc6"</span></span>
<span><span class="co">## LineB_G1 "4fc7b8af32ddd74e07cb49d147ef1938" "8967fabf10e55d881caa6fe192e7d4ca"</span></span>
<span><span class="co">##          1:5501-6500                       </span></span>
<span><span class="co">## LineA_G1 "1b568197f6f329ec5b71f66e49a732fb"</span></span>
<span><span class="co">## LineB_G1 "05efe15d97db33185b64821791b01b0f"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="haplotype-id-metadata">Haplotype ID metadata<a class="anchor" aria-label="anchor" href="#haplotype-id-metadata"></a>
</h3>
<p>To return metadata for each haplotype ID as a <code>tibble</code>
object, we can use the <code><a href="../reference/readHapIdMetaData.html">readHapIdMetaData()</a></code> function:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/readHapIdMetaData.html">readHapIdMetaData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 76 × 6</span></span></span>
<span><span class="co">##    hap_id                 sample_name description source checksum ref_range_hash</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 0eb9029f3896313aebc69… LineA       haplotype … /User… Md5      39f96726321b3…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 12f0cec9102e84a161866… LineA       haplotype … /User… Md5      546d1839623a5…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 13417ecbb38b9a159e3ca… LineA       haplotype … /User… Md5      5812acb1aff74…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 18498579d89483ac270e8… LineA       haplotype … /User… Md5      e07d04d2fc96b…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 184a72815a2ba5949635c… LineA       haplotype … /User… Md5      cb86faf105b19…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 1b568197f6f329ec5b71f… LineA       haplotype … /User… Md5      d896e9cc56e74…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 1e38bd82670c3f10982f7… LineA       haplotype … /User… Md5      db22dfc14799b…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 3149b3144f93134eb2966… LineA       haplotype … /User… Md5      57705b1e2541c…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 369464a8743d2e40ad83d… LineA       haplotype … /User… Md5      66465399052d8…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 3ec680649615da0685b8c… LineA       haplotype … /User… Md5      347f0478b1a55…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 66 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="haplotype-id-metadata-positions">Haplotype ID metadata (positions)<a class="anchor" aria-label="anchor" href="#haplotype-id-metadata-positions"></a>
</h3>
<p>To return positional information for each haplotype ID as another
<code>tibble</code> object, we can use the
<code><a href="../reference/readHapIdPosMetaData.html">readHapIdPosMetaData()</a></code> function:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/readHapIdPosMetaData.html">readHapIdPosMetaData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 76 × 5</span></span></span>
<span><span class="co">##    hap_id                           contig_start contig_end start   end</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 0eb9029f3896313aebc69c8489923141 2            2          <span style="text-decoration: underline;">49</span>501 <span style="text-decoration: underline;">50</span>300</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 12f0cec9102e84a161866e37072443b7 1            1              1  <span style="text-decoration: underline;">1</span>000</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 13417ecbb38b9a159e3ca8c9dade7088 2            2              1  <span style="text-decoration: underline;">1</span>000</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 18498579d89483ac270e8cca57f34752 1            1          <span style="text-decoration: underline;">16</span>501 <span style="text-decoration: underline;">17</span>500</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 184a72815a2ba5949635cc38769cedd0 2            2          <span style="text-decoration: underline;">12</span>001 <span style="text-decoration: underline;">16</span>500</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 1b568197f6f329ec5b71f66e49a732fb 1            1           <span style="text-decoration: underline;">5</span>501  <span style="text-decoration: underline;">6</span>500</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 1e38bd82670c3f10982f70390c599a8d 2            2          <span style="text-decoration: underline;">16</span>501 <span style="text-decoration: underline;">17</span>500</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 3149b3144f93134eb29661bade697fc6 1            1           <span style="text-decoration: underline;">1</span>001  <span style="text-decoration: underline;">5</span>500</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 369464a8743d2e40ad83d1375c196bdd 1            1           <span style="text-decoration: underline;">6</span>501 <span style="text-decoration: underline;">11</span>000</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 3ec680649615da0685b8c245e0f196e2 2            2          <span style="text-decoration: underline;">11</span>001 <span style="text-decoration: underline;">12</span>000</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 66 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="all-hvcf-data">All hVCF data<a class="anchor" aria-label="anchor" href="#all-hvcf-data"></a>
</h3>
<p>In a majority of cases, we may need more than one piece of hVCF data.
We can read all of the prior data simultaneously as
<code>PHGDataSet</code> object which is an in-R-memory representation of
all hVCF data:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/readPhgDataSet.html">readPhgDataSet</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## A <span style="font-weight: bold;">PHGDataSet</span> object</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of ref ranges....: <span style="font-weight: bold;">38</span></span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of samples.......: <span style="font-weight: bold;">2</span></span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of hap IDs.......: <span style="font-weight: bold;">76</span></span></span>
<span><span class="co">##  <span style="color: #00BB00;">❯</span> # of asm regions...: <span style="font-weight: bold;">76</span></span></span></code></pre>
<p>If this object is created, we can use the prior <code>read*</code>
methods to instantaneously pull out the previously mentioned R data
objects:</p>
<ul>
<li><code><a href="../reference/readSamples.html">readSamples()</a></code></li>
<li><code><a href="../reference/readRefRanges.html">readRefRanges()</a></code></li>
<li><code><a href="../reference/readHapIds.html">readHapIds()</a></code></li>
<li><code><a href="../reference/readHapIdMetaData.html">readHapIdMetaData()</a></code></li>
<li><code><a href="../reference/readHapIdPosMetaData.html">readHapIdPosMetaData()</a></code></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="filter-data">Filter data<a class="anchor" aria-label="anchor" href="#filter-data"></a>
</h2>
<p>In some cases, you may want to query information and focus on one
specific reference range and/or sample(s). We can
<strong>filter</strong> our <code>PHGDataSet</code> objects using the
<code>filter*</code> family of functions.</p>
<p>For the following examples, let’s picture our primary data as a
2-dimensional <code>matrix</code> of:</p>
<ul>
<li>samples (rows)</li>
<li>reference ranges (columns)</li>
<li>hap ID data (elements)</li>
</ul>
<p>If we were to represent this as an object in R called
<code>phgDs</code>, it would look like the following diagram:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phgDs</span></span></code></pre></div>
<p><img src="img/filtering/mat_01.svg" fig.align="center"></p>
<div class="section level3">
<h3 id="filter-by-sample">Filter by sample<a class="anchor" aria-label="anchor" href="#filter-by-sample"></a>
</h3>
<p>If we want to filter based on sample ID, we can use the
<code><a href="../reference/filterSamples.html">filterSamples()</a></code> function. Simply add the sample ID or
collection of sample IDs as a <code>character</code> string or a vector
of <code>character</code> strings, respectively. Using the prior diagram
as a reference, I will filter out anything that is not the
following:</p>
<ul>
<li><code>B73</code></li>
<li><code>Ky21</code></li>
<li><code>Mo17</code></li>
</ul>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phgDs</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/filterSamples.html">filterSamples</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B73"</span>, <span class="st">"Ky21"</span>, <span class="st">"Mo17"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/filtering/mat_02.svg" fig.align="center"></p>
<p><strong>Note</strong>: If samples are added to the filter collection,
but are not present in the data, they will be discarded. If
<strong>no</strong> samples are found, an exception will be thrown
stating that no samples could be found.</p>
</div>
<div class="section level3">
<h3 id="filter-by-reference-range">Filter by reference range<a class="anchor" aria-label="anchor" href="#filter-by-reference-range"></a>
</h3>
<p>If we want to filter based on sample ID, we can use the
<code><a href="../reference/filterRefRanges.html">filterRefRanges()</a></code> function. Currently, this takes a
<code>GRanges</code> object where we can specify integer-based ranges
located in genomic regions by chromosome (i.e., <code>seqnames</code>).
For example, if I want to return all ranges that intersect with the
following genomic regions:</p>
<table class="table">
<thead><tr class="header">
<th>Chromosome</th>
<th>Start (bp)</th>
<th>End (bp)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>"1"</code></td>
<td><code>100</code></td>
<td><code>400</code></td>
</tr>
<tr class="even">
<td><code>"2"</code></td>
<td><code>400</code></td>
<td><code>900</code></td>
</tr>
</tbody>
</table>
<p>I can create the following <code>GRanges</code> object and pass that
to the filter method:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">800</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">900</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">phgDs</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/filterRefRanges.html">filterRefRanges</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/filtering/mat_03.svg"></p>
</div>
<div class="section level3">
<h3 id="chaining-methods">Chaining methods<a class="anchor" aria-label="anchor" href="#chaining-methods"></a>
</h3>
<p>Since we have two dimensions, we can filter simultaneously by
“piping” or combining filter methods in one pass:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">800</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">900</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">phgDs</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterSamples.html">filterSamples</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B73"</span>, <span class="st">"Ky21"</span>, <span class="st">"Mo17"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/filterRefRanges.html">filterRefRanges</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<p><img src="img/filtering/mat_04.svg"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p><b>rPHG2</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
